src/main/java/gov/cabinetoffice/gapuserservice/service/user/OneLoginUserService.java

```diff
@@ -474,11 +475,38 @@ public boolean hasEmailChanged(final User user, final OneLoginUserInfoDto userIn
     }
 
     public List<UserEmailDto> getUserEmailsBySubs(List<String> subs) {
-        List<User> users = userRepository.findBySubIn(subs);
-        return users.stream().map(user -> UserEmailDto.builder()
-                .emailAddress(awsEncryptionService.encryptField(user.getEmailAddress()))
-                .sub(user.getSub())
-                .build())
+
+        // Splits the list of subs into two lists. One containing OneLogin subs and the other containing COLA subs.
+        final Map<Boolean, List<String>> partitionedSubs =
+                subs.stream()
+                    .collect(Collectors.partitioningBy(sub -> sub.contains(ONE_LOGIN_PREFIX)));
+
+        final List<User> users = new ArrayList<>();
+
+        if (!partitionedSubs.get(true).isEmpty()) {
+            final List<User> oneLoginUsers = userRepository.findBySubIn(partitionedSubs.get(true));
+            if (!oneLoginUsers.isEmpty()) {
+                users.addAll(oneLoginUsers);
+            }
+        }
+
+        if (!partitionedSubs.get(false).isEmpty()) {
+            final List<UUID> colaSubs = partitionedSubs.get(false).stream()
+                    .map(UUID::fromString)
+                    .toList();
+
+            final List<User> colaUsers = userRepository.findByColaSubIn(colaSubs);
+            if (!colaUsers.isEmpty()) {
+                users.addAll(colaUsers);
+            }
+        }
```

Comment:
Might be able to simplify this logic with the following JPA queries
```
    List<User> findBySubInAndSubStartsWith(Collection<String> subs, String sub);

    @Query("select u from User u where u.sub in ?1 and u.sub not like concat(?2, '%')")
    List<User> findBySubInAndSubNotStartsWith(Collection<String> subs, String sub);

Reply:
sub and cola sub are different cols in the database so would still need logic to partition the collection we've been sent anyway but I might be missing what you're saying here?